fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py

```diff
@@ -0,0 +1,828 @@
+                 action_config: dict):
+        self.node = node
+        self.name = name
+        self.api = mir_api
+        self.update_handle = update_handle
+        self.fleet_config = fleet_config
+        self.action_config = action_config
+
+
+class RobotAdapterMiR:
+    def __init__(
+        self,
+        name: str,
+        rmf_config: rmf_easy.RobotConfiguration,
+        mir_config: dict,
+        conversions: dict,
+        rmf_missions: dict,
+        fleet_handle,
+        fleet_config,
+        plugin_config: dict | None,
+        node: Node,
+        event_loop,
+        debug=False
+    ):
+        self.name = name
+        self.node = node
+        self.mission: MissionHandle | None = None
+        self.event_loop = event_loop
+        self.debug = debug
+
+        self.disconnect = False
+        self.disconnect_mutex = Lock()
+        self.replan_after_disconnect = False
+
+        prefix = mir_config['base_url']
+        headers = {
+            'Content-Type': mir_config['user'],
+            'Authorization': mir_config['password'],
+        }
+        self.api: MirAPI = MirAPI(prefix, headers, conversions, rmf_missions)
+
+        status = self.api.status_get()
+        while self.api.status_get() is None:
+            status = self.api.status_get()
+            time.sleep(0.1)
+        self.last_known_status: MirStatus = status
+        self.current_map = status.state.map
+        self.fleet_config = fleet_config
+
+        self.api.update_known_positions()
+        self.api.positions_delete()
+
+        # This must be done after the API is made
+        self.lift_positions: dict = conversions.get('lift_positions', {})
+        for lift, levels in self.lift_positions.items():
+            for level, position in levels.items():
+                position_name = position.get('name')
+                assert position_name is not None, (
+                    f'Missing "name" field for level [{level}] in lift '
+                    f'[{lift}] for lift_positions in configuration file'
+                )
+                known = self.api.known_positions.get(position_name)
+                assert known is not None, (
+                    f'MiR is missing the position [{position_name}] needed as '
+                    f'a lift position on level [{level}] for lift [{lift}]'
+                )
+        # We should only have one issue ticket at a time to manage unsuccessful
+        # navigation missions
+        self.nav_issue_ticket = None
+        self.requested_replan = False
+        self.replan_counts = 0
+
+        self.fleet_handle = fleet_handle
+        self.update_handle = fleet_handle.add_robot(
+            self.name,
+            status.state,
+            rmf_config,
+            self._make_callbacks(),
+        )
+
+        # Track the current ongoing action
+        self.current_action = None
+
+        # Import and store plugin actions and action factories
+        self.action_factories = {}
```

Comment:
I recommend iterating through the `plugin_config` and creating a dictionary that maps from action category to plugin name. Then when performing an action you would look up which plugin the user specified for that action and use that specific action factory instead of looping through all of them. This would help make sure we use the action factory that the user intended in situations where the same action could be performed by multiple different plugins. If the user specifies the same action category for multiple plugins then we would throw an exception right away.

We could still fall back on iterating through the action factories in situations where an incoming action category does not match any that were listed. That could happen if the user manually calls `add_performable_action` downstream.

Reply:
Gotcha, added the action-to-plugin dictionary in [8239b87](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/8239b87f58dbc0edad472705c1b8df7ecff89014).

> We could still fall back on iterating through the action factories in situations where an incoming action category does not match any that were listed. That could happen if the user manually calls add_performable_action downstream.

I removed the code to iterate through action factories as a fallback. I've been thinking about this and could not come up with a scenario where the user manually called `add_performable_action` downstream while the action config provides clue to which plugin this new action belongs to. If the action has already been provided in the action config, we would have added them as performable. If the action has not been provided, we would not know which plugin it is associated with.

Reply:
> If the action has not been provided, we would not know which plugin it is associated with.

The idea is that if a user manually calls `add_performable_action`, we can figure out which plugin to use by iterating through each factory and calling `factory.supports_action( _ )` to see which plugin supports the action.

That being said, this would be an extremely niche use case to the point that I wouldn't mind if we officially don't support it until someone asks for it with a compelling enough use case. It should be possible to add support for it in the future without breaking API, so we don't need to worry about this hypothetical right now.

Reply:
If you want to add these capabilities within this PR I have no objection, but I'm also perfectly happy to merge as-is and just leave an issue ticket for us to consider those cases in the future.